src/protocols/output_management.rs

```diff
@@ -0,0 +1,785 @@
+        zwlr_output_configuration_head_v1::{self, ZwlrOutputConfigurationHeadV1},
+        zwlr_output_configuration_v1::{self, ZwlrOutputConfigurationV1},
+        zwlr_output_head_v1::{self, AdaptiveSyncState, ZwlrOutputHeadV1},
+        zwlr_output_manager_v1::{self, ZwlrOutputManagerV1},
+        zwlr_output_mode_v1::{self, ZwlrOutputModeV1},
+    },
+    wayland_server::backend::ClientId,
+};
+
+const VERSION: u32 = 4;
+
+/*
+    <clientId<
+        <confID , IpcOutputMap>
+        <headId, String /* output name */>
+    >>
+*/
+
+#[derive(Debug)]
+pub struct OutputConfigurationData {
+    serial: u32,
+    new_out: IpcOutputMap,
+    heads: HashMap<ZwlrOutputConfigurationHeadV1, String /* output name */>,
+}
+
+#[derive(Debug)]
+struct ClientData {
+    heads: HashMap<String, (ZwlrOutputHeadV1, Vec<ZwlrOutputModeV1>)>,
+    // confs: HashMap<ZwlrOutputConfigurationV1, OutputConfigurationData>,
+    confs: HashMap<ZwlrOutputConfigurationV1, OutputConfigurationData>,
+    manager: ZwlrOutputManagerV1,
+}
+
+pub struct OutputManagementManagerState {
+    display: DisplayHandle,
+    serial: u32,
+    clients: HashMap<ClientId, ClientData>,
+    current_out: Option<IpcOutputMap>,
+}
+
+pub struct OutputManagementManagerGlobalData {
+    filter: Box<dyn for<'c> Fn(&'c Client) -> bool + Send + Sync>,
+}
+
+impl OutputManagementManagerState {
+    pub fn new<D, F>(display: &DisplayHandle, filter: F) -> Self
+    where
+        D: GlobalDispatch<ZwlrOutputManagerV1, OutputManagementManagerGlobalData>,
+        D: Dispatch<ZwlrOutputManagerV1, ()>,
+        D: Dispatch<ZwlrOutputConfigurationV1, ()>,
+        D: Dispatch<ZwlrOutputHeadV1, String>,
+        D: Dispatch<ZwlrOutputModeV1, ()>,
+        D: OutputManagementHandler,
+        D: 'static,
+        F: for<'c> Fn(&'c Client) -> bool + Send + Sync + 'static,
+    {
+        let global_data = OutputManagementManagerGlobalData {
+            filter: Box::new(filter),
+        };
+        display.create_global::<D, ZwlrOutputManagerV1, _>(VERSION, global_data);
+
+        Self {
+            display: display.clone(),
+            clients: HashMap::new(),
+            serial: 0,
+            current_out: None,
+        }
+    }
+}
+
+impl<D> GlobalDispatch<ZwlrOutputManagerV1, OutputManagementManagerGlobalData, D>
+    for OutputManagementManagerState
+where
+    D: GlobalDispatch<ZwlrOutputManagerV1, OutputManagementManagerGlobalData>,
+    D: Dispatch<ZwlrOutputManagerV1, ()>,
+    D: Dispatch<ZwlrOutputConfigurationV1, ()>,
+    D: Dispatch<ZwlrOutputHeadV1, String>,
+    D: Dispatch<ZwlrOutputModeV1, ()>,
+    D: OutputManagementHandler,
+{
+    fn bind(
+        state: &mut D,
+        display: &DisplayHandle,
+        client: &Client,
+        manager: New<ZwlrOutputManagerV1>,
+        _manager_state: &OutputManagementManagerGlobalData,
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        let manager = data_init.init(manager, ());
+        let outputs = state.get_outputs();
```

Comment:
The way I think this should work is:
* `OutputManagementManagerState` stores the "current state" as far as it knows. I suppose it already does, in `current_out`.
* When sending the state to the clients, it uses its stored "current state". Not the "newest state", to avoid any potential desync problems.
* The "current state" is updated to "newest state" only when niri calls `notify_changes`. This is also when state changes are pushed to the clients.

It already seems to work mostly like this, just need to replace some uses of "newest state" with the "current state".

Reply:
I see you reacted with +1 and re-requested a review, but I'm not sure this change has been completed yet, i.e. I think the `get_outputs()` call should be removed from the trait, and instead `current_out` should be used.

Reply:
So i gave the `outputs` at the initialization.